### PR TITLE
Ticket: 8331: Google Ads Report Data is Malformed for complex object

### PR DESCRIPTION
**Pull request by dmykhailov**
## What
Working with google ads I found an issue with one of the reports. We have report that is being generated from the following GAQL (number of selected columns is much bigger, but I have removed the most because they are not relevant):

```
SELECT
[ad_group_ad.ad.id](http://ad_group_ad.ad.id/),
ad_group_ad.resource_name,
ad_group_ad.ad.responsive_search_ad.headlines
FROM ad_group_ad
```

The problem that I am facing is with ad_group_ad.ad.responsive_search_ad.headlines column.Here is one sample of the values in the column:

```
[
"text: \"{KeyWord:Miro}\"\npinned_field: HEADLINE_1\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Comece uma retrospectiva\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Modelo de retrospectiva grátis\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Faça retrospectivas eficazes\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Faça retrospectivas na Miro\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Ouça o que o time tem a dizer\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Discuta e planeje melhorias\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Grátis, fácil, compartilhável\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Colabore em tempo real\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Mais de 300 modelos editáveis\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Mais de {CUSTOMIZER.User Amount} milhões de usuários\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n",
"text: \"Acesse em qualquer dispositivo\"\nasset_performance_label: PENDING\npolicy_summary_info {\n review_status: REVIEWED\n approval_status: APPROVED\n}\n"
]
```

So, it is list of strings and each string contains a representation of data which is not really JSON or yaml. that looks very similar to an string produced by protobuf toString() method.The problem is that it is quite hard to parse this values.

The type of the data from Google Ads GRPC response is "MESSAGE" which corresponds to the hypothesis of "toString()" method call.

## How
As a solution I would looks for converting GRPC message to JSON and storing JSON values for ease of integration and further parsing. 

## Review guide
That's a very simple fix that does not fail existing tests.

## User Impact
People who already relied on the text based parsing might get their code broken.

## Can this PR be safely reverted and rolled back?
- [X] YES 💚

## Help that I need
I would like to have some help with the testing approach. At this moment integration and unit tests are based on python object and the tests do not mimic GRPC messages, so, any edge cases from GRPC conversion world are not really covered.

### CODE REVIEW

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:173]: What is the exact exception type we expect back if this is not json serializable? For this `except` block we should try to target the specific exception that might be thrown so that we don't accidentally swallow unexpected ones

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:228]: This line may not have the right value you're expecting as input to `serialize_protobuf_message()`. This list comprehension will turn `Repeated`/`RepeatedComposite` into an underlying value within the object. So for this example `Repeated`:

[
  { "text": "An exciting headline", "policy_summary_info": { "approval_status": "APPROVED" } },
  { "text": "Two", "policy_summary_info": { "approval_status": "APPROVED" } }
]

What we end up passing to the method is the first `AdTextAsset`. So in that regard I'm not sure either the method signature, or what we pass your new method is right.

**Review by brianjlai — COMMENTED** [review]
Hi Dmytro. Thank you for the contribution!

I've added a couple of comments to integrate type checking more comprehensively into the work.

And although the change is straightforward, we should still be adding new unit tests to ensure that your changes have the intended outcomes. My suggestion is to add a new test to `test_google_ads.py`. There are some good examples to look from like `test_get_field_value()`, and I can also provide an example that I wrote:

```
def test_get_field_value_object():
    expected_response = {}
    field = "ad_group_ad.ad.responsive_search_ad.headlines"
    ads_row = GoogleAdsRow(
        ad_group_ad={
            "ad": {
                "responsive_search_ad": {
                    "headlines": [
                        {
                            "text": "An exciting headline",
                            "policy_summary_info": {
                                "review_status": "REVIEWED",
                                "approval_status": "APPROVED"
                            }
                        },
                        {
                            "text": "second"
                        }
                    ]
                }
            }
        }
    )

    response = GoogleAds.get_field_value(ads_row, field, {})
    assert response == expected_response
```

I think the main issue I noticed while I was reviewing the code is that we're not properly parsing the incoming `RepeatedComposite`. Using the above code we'll end up with `AttributeError: Unknown field for AdTextAsset: pb`. It's possible that my test case isn't testing exactly the issue you've highlighted, but let's try to figure out what the ideal test case is and then we can figure out how to implement it

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (dmykhailov) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py]: I've added the type hint. Here it will be "Message" from protobuf which expectedly be the parent for Google Ads entities.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (dmykhailov) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py]: Because we are passing not an instance of a Repeated or RepeatedComposite, but whatever is repeated items inside those are, I cannot read the public Pb property of those.

Unless we would consider serialization of the entire object without drilling into its elements.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (dmykhailov) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py]: I've added the type of AttributeError cause that is what I expect. If there is no "_pb" then fallback to str().

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (dmykhailov) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:233]: Each individual element of the `Repeated` or `RepeatedCoimposite` is indeed passed to the serialization. I'l explore if the whole object serialization will pass the tests.

Comment (dmykhailov) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:233]: I've just tried to serialize `Repeated` value via the same method using the public 'pb' and got an error that the object does not have "DESCRIPTOR" attribute. If I got it right, it means that the object is not protobuf message compatible with the serializer.

**Comment by dmykhailov** [comment]
@brianjlai thank you for your review and feedback. I've added two tests inspired by the example that you provided. Also, added a bit of comments and type hints where needed.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:174]: nit change `cab` to `can`

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:176]: Can we add `# type: ignore[union-attr] # if _pb does not exist then we fall back to using the string representation` . We plan to start using mypy to enforce type safety and i have a feeling accessing a potentially non-existent field will cause it to fail

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (brianjlai) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:233]: ah i see. It feels very weird that we would not end up with a protobuf message at all. What we may want to do instead of relying solely on the try/except block, we should add an explicit type check `isinstance()` and if it not of the correct type, then we defer to `str()`. We can still leave the try catch, but I think proactively checking the type and using str() when its the wrong type is a better pattern

one more question around this. have you been able to test the input for your specific use case? Our tests are semi-stripped down mocks of a real AdRow, before we do final testing and merge, it would be great to verify that given inputs similar to what you have, that we do end up serializing this correctly

**Review by brianjlai — COMMENTED** [review]
a few additional question and small nits. Thanks for the latest updates

**Comment by dmykhailov** [comment]
@brianjlai yes. I was able to test exactly my case with integration tests using our query. I haven't added that to the code cause the query is defined in `secrets/config.json`.

But also, the unit tests that we added are actually cover the headlines field that should be fixed. If I remove the fix from the code and re-run the unit tests it shows exactly the behavior that should be fixed and the test is failing. It looks to me that it is a good approach build the objects are `GoogleAdsRow` because during creation it uses protobuf mechanics.

**Comment by igorlushchyk** [comment]
Really useful PR for us!

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py
Comment (brianjlai) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py:177]: the mypy convention over ignore the type check should actually go on the line of code what has the type mismatch, so it should look like:

```
return json_format.MessageToJson(message._pb, indent=0).replace('\n', '')  # type: ignore[union-attr] # add explanation for why here
```

**Review by brianjlai — COMMENTED** [review]
> @brianjlai yes. I was able to test exactly my case with integration tests using our query. I haven't added that to the code cause the query is defined in secrets/config.json.

@dmykhailov Thanks for confirming

Things look good overall, but I do see that we have some code formatting errors that are causing us not to run the full test suite. Have you run the formatting checks via airbyte-ci (`airbyte-ci format fix all`) which should automatically fix your branch. And then you'll need to push one more commit.

Once we get that done, we can see how our automated test suite goes before approving. And at the same time, I will take your code and run some of our regression tests against it and let you know how that goes. I'll try to find time to do that early next week, but it might not be right at the start. Thank you in advance for your patience

**Comment by dmykhailov** [comment]
@brianjlai the formatting issues has been fixed. Thanks for point out the `airbyte-ci` tool.

**Comment by brianjlai** [comment]
@dmykhailov thanks for the updates. I think the change looks good, but I'm not sure if we need to run something to trigger the community PR test workflow. @marcosmarxm mind lending a had here. While we're getting CI to run on this, I'll try to do some additional functional testing on my side before approving and that should hopefully get us to the finish line

**Comment by brianjlai** [comment]
@dmykhailov thank you for the patience. I did some functional testing and running this locally to validate the expected behavior. Things look good overall, but just wanted to confirm that the expected shape of the data you were also seeing in the output records looked something like this:

```
"ad_group_ad.ad.responsive_display_ad.headlines": [
        "{\"text\": \"An exciting headline\",\"policySummaryInfo\": {\"reviewStatus\": \"REVIEWED\",\"approvalStatus\": \"APPROVED\"}}",
        "{\"text\": \"An exciting headline\",\"policySummaryInfo\": {\"reviewStatus\": \"REVIEWED\",\"approvalStatus\": \"APPROVED\"}}",
        "{\"text\": \"An exciting headline\",\"policySummaryInfo\": {\"reviewStatus\": \"REVIEWED\",\"approvalStatus\": \"APPROVED\"}}"
      ],
```

If we're on the same page, we can move forward with bumping the version and updating the changelog before final approval and merge. For an example of the various files to update look at this example https://github.com/airbytehq/airbyte/pull/44016. Specifically the `pyrproject.yaml`, `metadata.yaml`, and `posthog.md` (for you it would be google-ads.md

**Comment by dmykhailov** [comment]
@brianjlai yes. that is the expected output. 

Should I do the version bump or you'll do that?

**Comment by dmykhailov** [comment]
@brianjlai I've updated the mentioned files to the best of my knowledge. Please review and let me know if I missing something.

**Comment by brianjlai** [comment]
> Should I do the version bump or you'll do that?

I see you've done that, and it's probably easier if you did. We want the version increments to be in the same PR that the change is in for better keeping track of things. Thanks for doing that.

Let's see about getting a new CI run of this kicked off

**Comment by brianjlai** [comment]
Okay, the last CI run has a few docs formatting failures, but it's unrelated to your change. The change looks good to me and all set to merge. Thanks for the contribution @dmykhailov 

@natikgadzhi we have the unrelated failing QA check due to docs formatting. We can fix that on our own time and it should not need to block @dmykhailov 's contribution. Can you run the approve override to release this change when you have a moment. And I'll monitor things as the connector get's published

**Comment by natikgadzhi** [comment]
Sure! I'll publish in the morning so you don't have to stress at 9pm ;-)

**Comment by dmykhailov** [comment]
@brianjlai @natikgadzhi I've just resolved the merge conflict and bumped the version of this update accordingly.

**Review by natikgadzhi — APPROVED** [review]

**Pull request merged**
